### ISSUE TITLE
Feature Request: bfloat16 training

### ISSUE THREAD

**Issue by ikergarcia1996** [huggingface/accelerate#243]
Hi!
Thanks for the accelerate project, It is helping me a lot in my research :)

I am working with T5 models. Training T5 models (I found this issue with mT5 and T5-large) with FP16 is not possible (without tricks that may slowdown the training) the loss will overflow or underflow leading to NaN loss (https://github.com/huggingface/transformers/pull/10956). The best solution proposed up to date is training with BF16 instead of FP16. However I cannot find any option to use BF16 instead of FP16 in accelerate. 
![LossNan](https://user-images.githubusercontent.com/18737249/151950811-939f9bc1-6ce2-446b-b76f-8743c8f4a98e.png)

The adoption of the bf16 format is spreading and more and more models are being trained in this format. In addition, hardware that supports bf16 is also becoming more widespread. Would it be possible to add support for bf16 in accelerate? Or is there a way to use BF16 that I haven't figured out? I can't find anything in the documentation

**Comment by sgugger** [comment]
Hi there! You're right it's not added yet. It came up pretty recently (and you will need a very recent GPU) or were you talking of bfloat16 support for TPU training?

**Comment by ikergarcia1996** [comment]
I am talking about running bfloat16 in a GPU. I am currently experimenting with an Ampere GPU that supports this feature. 
Taking a look at the AMP documentation in PyTorch, using bf16 should be as simple as adding an additional parameter to autocast 
```
with autocast(dtype=torch.bfloat16):
    ...
```
Would it be possible to add an option to the configuration to specify the data type of the autocast function? 
I could help with a pull request if that's OK.

**Comment by sgugger** [comment]
You just have to be a bit careful since that argument was only introduced in recent versions of PyTorch and Accelerate supports version 1.4.0 and above, but I think that's the gist of it yes. I'm happy to review a PR if you want to dive into this :-)

**Comment by neel04** [comment]
>... bfloat16 support for TPU training?

Just to confirm - TPU training uses bfloat16 right? And does it work in multi-TPU settings (as provided by [TRC)](https://sites.research.google/trc/about/)

**Comment by sgugger** [comment]
They use bfloat16 internally, but convert back outputs of the layers to FP32. There is a an environment variable you can set to have the training in full bfloat16.

Closing the issue since the PR adding support on GPU has been merged :-)

**Comment by cyk1337** [comment]
Hi @ikergarcia1996, I also met the same `Nan` problem for training mt5 using fp16. How did you solve it?

**Comment by ikergarcia1996** [comment]
@cyk1337 You must train the model using BF16 instead of FP16 (Accelerate already supports bf16). If you GPU doesn't support BF16 you should train the model using FP32. I don't know if somebody has find a way to successfully finetune T5 in FP16.

**Comment by lixin4ever** [comment]
@ikergarcia1996 BF16 indeed works and no NAN problem occurs. Just wonder did you observe any performance drop after using BF16? In our experiment, we find that the performance of mT5 trained with BF16 is pretty much worse than the mT5 trained with FP32
